src/ir/properties.h

```diff
@@ -335,6 +335,51 @@ inline bool canEmitSelectWithArms(Expression* ifTrue, Expression* ifFalse) {
   return ifTrue->type.isSingle() && ifFalse->type.isSingle();
 }
 
+// An intrinsically-nondeterministic expression is one that can return different
+// results for the same inputs, and that difference is *not* explained by
+// other expressions that interact with this one. Hence the cause of
+// nondeterminism can be said to be "intrinsic" - it is internal and inherent in
+// the expression.
+//
+// To see the issue more concretely, consider these:
+//
+//    x = load(100);
+//    ..
+//    y = load(100);
+//
+//  versus
+//
+//    x = struct.new();
+//    ..
+//    y = struct.new();
+//
+// Are x and y identical in both cases? For loads, we can look at the code
+// in ".." to see: if there are no possible stores to memory, then the
+// result is identical (and we have EffectAnalyzer for that). For the GC
+// allocations, though, it doesn't matter what is in "..": there is nothing
+// in the wasm that we can check to find out if the results are the same or
+// not. (In fact, in this case they are always not the same.) So the
+// nondeterminism is "intrinsic."
+//
+// Thus, loads are nondeterministic but not intrinsically so, while GC
+// allocations are actual examples of intrinsically nondeterministic
+// instructions. If wasm were to add "get current time" or "get a random number"
+// instructions then those would also be intrinsically nondeterministic.
+//
+//  * Note that NaN nondeterminism is ignored here. Technically that allows e.g.
+//    an f32.add to be nondeterministic, but it is a valid wasm implementation
+//    to have deterministic NaN behavior, and we optimize under that assumption.
+//    So NaN nondeterminism does not cause anything to be intrinsically
+//    nondeterministic.
+//  * Note that calls are ignored here. In theory this concept could be defined
+//    either way for them (that is, we could potentially define them as either
+//    intrinsically nondeterministic, or not, and each could make sense in its
+//    own way). It is simpler to ignore them here, which means we only consider
+//    the behavior of the expression provided here (including its chldren), and
+//    not external code.
+//
+bool isIntrinsicallyNondeterministic(Expression* curr, FeatureSet features);
```

Comment:
Drive-by comment: Perhaps "generative" would be a better term here than "intrinsically nondeterministic," since these instructions always generate new values but aren't really nondeterministic in the normal sense - they generate deterministically fresh values with deterministic structure. This would also be shorter and allow the discussion of NaNs to be elided because floating point operations are clearly not generative.

Reply:
I agree "generative" would capture the concept of allocation, but if wasm adds a "get current time" or "get a random number" methods then we'd need something more than just allocation here. I'm open to better terms than nondeterminism, though - I considered "nonconstant", "impure", but couldn't think of anything good...

Reply:
(I'll land this once tests pass, to unblock j2cl opts that depend on this, but let's keep discussing the terminology.)

Reply:
I think "generative" would cover something like getting the current time, too; each time the instruction is executed it (potentially) generates a fresh value.

Reply:
Interesting. Yes, I can see how that works here. And for random numbers, again, it is generating a new value...

I don't have a strong preference myself either way, I think. So happy to go with yours if you feel strongly, and depending on what @aheejin thinks?

Reply:
I would say I feel medium-strongly about it :) Also would be happy to hear what @aheejin thinks.

Reply:
I think "generative" sounds good, but I'm OK with either. But even if we pick "generative", it'd be better to have clarifying comments on what "generative" means here and the difference between things like NaN and these.